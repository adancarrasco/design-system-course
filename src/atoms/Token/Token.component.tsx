import * as React from 'react'
import styled from '@emotion/styled'

export interface ITokenProps extends ITokenStyledProps {
  children: React.ReactNode | string
}

const Token: React.FC<ITokenProps> = ({ children, ...rest }) => (
  <TokenStyled role="button" {...rest}>
    {children}
  </TokenStyled>
)

export interface ITokenStyledProps {
  borderRadius?: string
  borderColor?: string
  backgroundColor?: string
  color?: string
  isActive?: boolean
  activeBorderColor?: string
  activeBackgroundColor?: string
  activeColor?: string
}

const TokenStyled = styled.span<ITokenStyledProps>`
  cursor: pointer;
  background-color: ${({ isActive, backgroundColor, activeBackgroundColor }) =>
    isActive ? activeBackgroundColor ?? '#c6dff5' : backgroundColor ?? '#fff'};
  border: 1px solid;
  border-color: ${({ isActive, borderColor, activeBorderColor }) =>
    isActive ? activeBorderColor ?? '#5490ea' : borderColor ?? '#ababab'};
  border-radius: ${(props) => props.borderRadius ?? '4px'};
  color: ${({ isActive, color, activeColor }) =>
    isActive ? activeColor ?? '#2763bd' : color ?? '#000'};
  display: inline-flex;
  font-size: 0.8rem;
  font-weight: ${(props) => (props.isActive ? 'bold' : 'normal')};
  padding: 0.2rem 0.3rem;
`

export default Token
