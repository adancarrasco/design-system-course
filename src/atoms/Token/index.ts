export { default } from './Token.component'

export type { ITokenProps, ITokenStyledProps } from './Token.component'
