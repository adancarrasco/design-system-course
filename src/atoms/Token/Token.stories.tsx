import React from 'react'
import { storiesOf } from '@storybook/react'
import { AccessAlarm } from '@material-ui/icons'

import Token from '.'

storiesOf('Atoms|Token', module)
  .add('Default', () => <Token>Token</Token>)
  .add('With custom props', () => (
    <Token borderColor="#aa0000" borderRadius="12px" backgroundColor="#fefefe" color="#aa0000">
      My token
    </Token>
  ))
  .add('With icon', () => (
    <Token>
      <AccessAlarm />
    </Token>
  ))
  .add('Active', () => <Token isActive>Token</Token>)
