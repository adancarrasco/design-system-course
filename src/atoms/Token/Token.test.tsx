import React from 'react'
import { render } from '@testing-library/react'
import { matchers } from 'jest-emotion'
import '@testing-library/jest-dom/extend-expect'

import Token, { ITokenProps } from '.'

expect.extend(matchers)

describe('Token Component', () => {
  const renderComponent = (props: ITokenProps) => render(<Token {...props}>{props.children}</Token>)

  it('matches snapshot and default render', () => {
    const props = {
      children: 'Token',
    }
    const { getByText } = renderComponent(props)
    const container = getByText(props.children)
    expect(container).toMatchSnapshot()
    expect(container).toBeInTheDocument()
  })

  it('should render with custom props', () => {
    const props = {
      children: 'Token',
      color: 'red',
      backgroundColor: 'blue',
      borderColor: 'red',
    }
    const { getByText } = renderComponent(props)
    const container = getByText(props.children)
    expect(container).toHaveStyleRule('color', 'red')
    expect(container).toHaveStyleRule('background-color', 'blue')
    expect(container).toHaveStyleRule('border-color', 'red')
  })

  it('should render active state', () => {
    const props = {
      children: 'Token',
      isActive: true,
    }
    const { getByText } = renderComponent(props)
    const container = getByText(props.children)
    expect(container).toHaveStyleRule('border-color', '#5490ea')
    expect(container).toHaveStyleRule('background-color', '#c6dff5')
    expect(container).toHaveStyleRule('color', '#2763bd')
    expect(container).toHaveStyleRule('font-weight', 'bold')
  })
})
