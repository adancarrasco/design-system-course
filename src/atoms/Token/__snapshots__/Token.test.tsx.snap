// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Token Component matches snapshot and default render 1`] = `
.emotion-0 {
  cursor: pointer;
  background-color: #fff;
  border: 1px solid;
  border-color: #ababab;
  border-radius: 4px;
  color: #000;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 0.8rem;
  font-weight: normal;
  padding: 0.2rem 0.3rem;
}

<span
  class="emotion-0"
  role="button"
>
  Token
</span>
`;
